ott/core/layers.py

```diff
@@ -0,0 +1,136 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Layers used in input convex neural networks (Amos+(2017), Bunne+(2022))."""
+
+from typing import Any, Callable, Tuple
+
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+
+PRNGKey = Any
+Shape = Tuple[int]
+Dtype = Any
+Array = Any
+
+
+class PositiveDense(nn.Module):
+  """A linear transformation using a weight matrix with all entries positive.
+
+  Args:
+    dim_hidden: the number of output dim_hidden.
+    rectifier_fn: choice of rectiver function (default: softplus function).
+    inv_rectifier_fn: choice of inverse rectiver function
+      (default: inverse softplus function).
+    beta: inverse temperature parameter (default: 1.0)
+    dtype: the dtype of the computation (default: float32).
+    precision: numerical precision of computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrix.
+    bias_init: initializer function for the bias.
+  """
+  dim_hidden: int
+  rectifier_fn: Callable = nn.softplus
+  inv_rectifier_fn: Callable = lambda x: jnp.log(jnp.exp(x) - 1)
```

Comment:
i think the assert that the rectifier and inv_rectifier match is missing

Reply:
Moved it.